contributors/design-proposals/pod-priority-resourcequota.md

```diff
@@ -75,5 +75,51 @@ Unlike pod spec, ResourceQuota does not have Priority `*int32` in its spec. As e
 ### Expected behavior of ResourceQuota admission controller
 
 * We do not enforce referential integrity across objects. i.e. Creation or update of `ResourceQuota` object which new `ResourceQuota` names a `PriorityClass` that does not exist, and empty `PriorityClassNameList` field are all allowed
-* It is configurable for `ResourceQuota` admission controller to reject pod with priority set but has no matched ResourceQuota found or not.ß
 * `ResourceQuota` admission controller **should not reject** deletion of `ResourceQuota` objects with reference.
+
+#### Changes in AdmissionConfiguration
+
+It is configurable for `ResourceQuota` admission controller to reject pod with priority set but has no matched `ResourceQuota` found or not. 
+
+This requires an extra boolean option in `LimitedResource` type:
+```go
+// LimitedResource matches a resource whose consumption is limited by default.
+// To consume the resource, there must exist an associated quota that limits
+// its consumption.
+type LimitedResource struct {
+  APIGroup string `json:"apiGroup,omitempty"`
+
+  Resource string `json:"resource"`
+
+  MatchContains []string
+
+  // AllowPriorityUnbound defines whether admission controller will or
+  // will not reject object with priority been set but have no matched
+  // resource quota.
+  // For example, if set to true, a pod with priority set but ResourceQuota
+  // is unbounded will be allowed.
+  // Default is true.
+  AllowPriorityUnbound bool `json:"allowpriorityunbound,omitempty"`
```

Comment:
If we decided to add this flag, we should find a better name. Something like, 'DefaultUnbound'.

Reply:
Good point!

Reply:
Booleans tend to proliferate as additional "modes" are later introduced for the same overall condition. So, in this context, if priority is assigned but no quota match exists, a boolean restricts us to two well defined but very inflexible behaviors (accept/reject). If we later come up with some other behavior we'd like to support, we'd have to either introduce more booleans or redefine the behavior of the existing boolean. The former quickly becomes incomprehensible to users (to figure out the boolean combinations to get the desired overall "mode") and the latter is behavioral API breakage.

Another possible way to express the configuration would be to define something like:

```go
type LimitedResource struct {
    UnmatchedQuotaPolicy string
}
```

Where `UnmatchedQuotaPolicy` is one of:

```go
const (
    AllowUnmatchedQuotaIfPriorityPresent = "AllowUnmatchedQuotaIfPriorityPresent"
    RejectUnmatchedQuotaIfPriorityPresent = "RejectUnmatchedQuotaIfPriorityPresent"
)
```

The "policy" approach seems to be more explicit and clear in terms of expressing the configuration while also leaving a lot of space for additional behaviors as other use cases are identified.

Just a thought. What do you all think?

Reply:
I like what @ironcladlou suggested and agree that in the future more options may be needed. The only thing that I would change is that after the changes in this proposal are implemented, priority will always be implied in ResourceQuota and also during Pod creation. So, I would change the names of unmatchedQuotaPolicy to something like:
AdmitIfNoMatchingResourceQuotaExists and
RejectIfNoMatchingResourceQuotaExists

Reply:
Like this idea too, thanks!

Reply:
I think the global configuration should at least match what the per namespace configuration allows.

In particular, I want to be able to say "in order to use priority: cluster-services, then you must have explicit quota".  To use any lower priority or other priority, I don't care if you have explicit quota.  So the bool, or even the enum string is not sufficiently expressive.

*Example: Any pod with priority cluster-services, or other-important-stuff must require explicit quota*
```yaml
apiVersion: apiserver.k8s.io/v1alpha1
kind: AdmissionConfiguration
plugins:
- name: "ResourceQuota"
  configuration:
  apiVersion: resourcequota.admission.k8s.io/v1alpha1
  kind: Configuration
  limitedResources:
  - resource: pods
    matchPriorityName:
    - "cluster-services"
    - "other-important-stuff"
```

*Example: Any pod with priority set must require explicit quota*
```yaml
apiVersion: apiserver.k8s.io/v1alpha1
kind: AdmissionConfiguration
plugins:
- name: "ResourceQuota"
  configuration:
  apiVersion: resourcequota.admission.k8s.io/v1alpha1
  kind: Configuration
  limitedResources:
  - resource: pods
    matchPriorityName:
    - "*"
```

Reply:
or just a syntax that let us say "pod with priority range 50-100" must have explicit quota would also work.

Reply:
@derekwaynecarr good points... your proposal looks best to me so far

Reply:
Based on what we discussed and decided in the meeting today, we are not going to support priority ranges in quota. ResourceQuota objects can have a list of PriorityClass names (or a wildcard) and they match against those PriorityClasses of quotas.

> In particular, I want to be able to say "in order to use priority: cluster-services, then you must have explicit quota". To use any lower priority or other priority, I don't care if you have explicit quota. So the bool, or even the enum string is not sufficiently expressive.

This is a good point and many customers will probably need it. I think a global config that specifies "whether lack of ResourceQuota should be considered unbound quota or no quota" can address this problem. Let's call it `the flag` in this comment. Admin can set the flag so that lack of RQ is unbound quota. Admin then creates RQ that matches `cluster-services` priority. Pods with other priority classes are unbound, but pods with `cluster-services` can consume as much as the RQ allows.
One limitation of this approach is that admin won't be able to set `the flag` per namespace. If this limitation is a concern, we can make `the flag` a namespace config. When `the flag` is not specified, the default behavior is that lack of RQ means no quota for restricted resources. Non-restricted resources will always be unbound.

Reply:
Thanks for quick update. Summary from my sides is (correct me please):
1. According to https://github.com/kubernetes/community/pull/933/files#r146907636, the default behavior is: if I have no quota, then I can consume an unbounded amount of the resource.
2. We need to make 1) configurable (unbounded amount VS no resource to use). The approaches are global config VS namespaced config.

For 2), it seems for if using namespaced config, users are expected to create a `ConfigMap`, right? Maybe a global config is enough?

Reply:
kindly ping @bsalamat to review the previous summary comment :)

Reply:
> According to https://github.com/kubernetes/community/pull/933/files#r146907636, the default behavior is: if I have no quota, then I can consume an unbounded amount of the resource.
We need to make 1) configurable (unbounded amount VS no resource to use). The approaches are global config VS namespaced config.
For 2), it seems if using namespaced config, users are expected to create a ConfigMap, right? Maybe a global config is enough?

I am fine with this approach and a global config. Let's see what @derekwaynecarr thinks.

Reply:
Kindly ping @derekwaynecarr , do you think this summary of design make sense?